### ISSUE TITLE
TRITON server returning wrong inferences

### ISSUE THREAD

**Issue by audrey-siqueira** [triton-inference-server/server#3325]
**Description**
After a while struggling with the TRITON server to be able to make inferences using my SavedModel (.pb), I finally managed to make the inferences by adding the parameter **--backend-config=tensorflow,version=2** to launch TRITON, maybe just in this  way was possible because my model was created in tensorflow=2.5.

The problem now is that I'm getting inferences with wrong results, I've already checked my Labels.txt document to confirm that I'm not making a mistake in the order of the labels.

Does anyone know why I might be getting wrong inferences? I am using the **image_client.py** client script.

**Triton Information**
Container  21.08

Are you using the Triton container or did you build it yourself? Triton Container

**To Reproduce**
My Model structure:

{
"name": "clasificacion_categorica",
"platform": "tensorflow_savedmodel",
"backend": "tensorflow",
"version_policy": {"latest": {"num_versions": 1}
},
"max_batch_size": 1,
"input": [
{
"name": "input_1",
"data_type": "TYPE_FP32",
"dims": [
256,
256,
3
]
}
],
"output": [
{
"name": "dense",
"data_type": "TYPE_FP32",
"dims": [
11
]
}
],
"batch_input": [],
"batch_output": [],
"optimization": {
"priority": "PRIORITY_DEFAULT",
"input_pinned_memory": {
"enable": true
},
"output_pinned_memory": {
"enable": true
}
},
"instance_group": [
{
"name": "clasificacion_categorica",
"kind": "KIND_GPU",
"count": 1,
"gpus": [
0
],
"profile": []
}
],
"default_model_filename": "model.savedmodel",
"cc_model_filenames": {},
"metric_tags": {},
"parameters": {},
"model_warmup": []
}

I don't understand why I get the wrong inferences, when I use the model with tensorflow serving, I get the correct results of the inferences, but using TRITON I get wrong results.

**Comment by dzier** [comment]
Can you please attach a detailed Triton log and include the command used to launch Triton.

**Comment by audrey-siqueira** [comment]
Hi @dzier , as my data is confidential, I'm going to use here a respective classification model (CNN) between cats and dogs which is resulting in the same problem.

**The command to launch TRITON:** sudo docker run --gpus=1 --rm -p8000:8000 -p8001:8001 -p8002:8002
 -v/home/audrey_siqueira/linux/models:/models  nvcr.io/nvidia/tritonserver :21.08-py3 tritonserver 
--model-repository=/models
 --backend-config=tensorflow,version=2

**The structure of the config.pbtxt file:**

    name: "CNN"
    platform: "tensorflow_savedmodel"
    backend: "tensorflow"
    version_policy: {latest: {num_versions: 1}}
    max_batch_size: 1

    input: [
        {
            name: "conv2d_input"
            data_type: TYPE_FP32
            format: FORMAT_NHWC
            dims: [64.64.3]
        }
    ]

    output: [
        {
            name: "dense_1"
            data_type: TYPE_FP32
            dims: [1]
            label_filename: "CNN_labels.txt"
        }
    ]

    batch_input: []
    batch_output: []

    optimization: {
        priority: PRIORITY_DEFAULT
        input_pinned_memory: {enable: true }
        output_pinned_memory: {enable: true}
        gather_kernel_buffer_threshold: 0
        eager_batching: false
    }

    instance_group: [
        {
            name: "CNN"
            kind: KIND_GPU
            count: 1
            gpus: [0]

            secondary_devices: []
            profile: []
            passive: false
            host_policy: ""
        }
    ]

    default_model_filename: "model.savedmodel"
    cc_model_filenames: {}
    metric_tags: {}
    parameters: {}
    model_warmup: []

**ps: My label.txt file contains Dog and Cat respectively.**


The model loads **correctly**, and I use the following command to make the inference:

**python image_client.py -m CNN -c 1 -s INCEPTION animals**

In which animals is a folder containing images of dogs and cats.

**The problem comes now, the output of my inference of 18 requests results in the following:**

Request 1, batch size 1
0.000000 (0) = Dog
Request 2, batch size 1
0.000001 (0) = Dog
Request 3, batch size 1
0.000000 (0) = Dog
Request 4, batch size 1
0.000102 (0) = Dog
Request 5, batch size 1
0.000000 (0) = Dog
Request 6, batch size 1
0.967779 (0) = Dog
Request 7, batch size 1
0.000000 (0) = Dog
Request 8, batch size 1
0.000000 (0) = Dog
Request 9, batch size 1
0.974457 (0) = Dog
Request 10, batch size 1
0.302763 (0) = Dog
Request 11, batch size 1
0.047200 (0) = Dog
Request 12, batch size 1
0.000063 (0) = Dog
Request 13, batch size 1
0.002313 (0) = Dog
Request 14, batch size 1
0.012119 (0) = Dog
Request 15, batch size 1
0.977927 (0) = Dog
Request 16, batch size 1
0.000000 (0) = Dog
Request 17, batch size 1
0.990310 (0) = Dog
Request 18, batch size 1
0.999618 (0) = Dog
PASS



It seems that the inference is not being made, and since I don't get any error log I don't understand why it's returning these inferences to me, the correct thing would be half of the inferences being cat and the other half being dog.

Someone has any ideia how to solve this?

**Comment by deadeyegoodwin** [comment]
What is the content of CNN_labels.txt?
What if you run with "-c 2"
You can see that you are getting different probability values...
Are you sure that INCEPTION is the correct scaling for your model? Does it match what you used during training?

**Comment by audrey-siqueira** [comment]
Hi, thanks for the reply.

Which alternatives exist beside INCEPTION? Maybe that would be the problem, I just know the INCEPTION command in TRITON.

I didn't train the model, a person at work just sent me the .pb model

If I change it to -c 2 , what happens is that the first probability is  always 1 and all others to are 0.

Which commands could I use besides INCEPTION?

**Comment by dzier** [comment]
Can you look in the `image_client.py` script to see what other options exist for `-s`?

**Comment by deadeyegoodwin** [comment]
image_client.py is just an example... you should look at it and modify it as necessary to match the requirements of your model.

**Comment by audrey-siqueira** [comment]
Hi, thanks for the answers.

I found the correct solution by changing the following parameter in image.client.py :

I just changed this 

**if scaling == 'INCEPTION':
         scaled = (typed / 127.5) - 1**

for this 

**if scaling == 'INCEPTION':
         scaled = (typed / 255)**


Your answers were very efficient to find the solution, thank you and I think you can close the issue.
